src/raftstore/store/snap.rs

```diff
@@ -61,223 +108,525 @@ impl Display for SnapKey {
+
+    pub const CRC32_BYTES_COUNT: usize = 4;
+    const DEFAULT_READ_BUFFER_SIZE: usize = 4096;
+
+
+    /// A structure represents the snapshot.
+    ///
+    /// All changes to the file will be written to `tmp_file` first, and use
+    /// `save` method to make them persistent. When saving a crc32 checksum
+    /// will be appended to the file end automatically.
+    pub struct Snap {
+        file: PathBuf,
+        digest: Option<Digest>,
+        size_track: Arc<RwLock<u64>>,
+        // File is the file obj represent the tmpfile, string is the actual path to
+        // tmpfile.
+        tmp_file: Option<(File, String)>,
+        reader: Option<File>,
+    }
+
+    impl Snap {
+        pub fn new_to_write<T: Into<PathBuf>>(dir: T,
+                                              size_track: Arc<RwLock<u64>>,
+                                              is_sending: bool,
+                                              key: &SnapKey)
+                                              -> io::Result<Snap> {
+            let mut path = dir.into();
+            try!(Snap::prepare_path(&mut path, is_sending, key));
+
+            let mut f = Snap {
+                file: path,
+                digest: Some(Digest::new(crc32::IEEE)),
+                size_track: size_track,
+                tmp_file: None,
+                reader: None,
+            };
+            try!(f.init_to_write());
+            Ok(f)
+        }
 
-    /// Get a validation reader.
-    pub fn reader(&self) -> io::Result<SnapValidationReader> {
-        SnapValidationReader::open(self.path())
-    }
+        pub fn new_to_read<T: Into<PathBuf>>(dir: T,
+                                             size_track: Arc<RwLock<u64>>,
+                                             is_sending: bool,
+                                             key: &SnapKey)
+                                             -> io::Result<Snap> {
+            let mut path = dir.into();
+            try!(Snap::prepare_path(&mut path, is_sending, key));
+
+            let reader = try!(File::open(&path));
+            let f = Snap {
+                file: path,
+                digest: None,
+                size_track: size_track,
+                tmp_file: None,
+                reader: Some(reader),
+            };
+            Ok(f)
+        }
 
-    pub fn exists(&self) -> bool {
-        self.file.exists() && self.file.is_file()
-    }
+        fn prepare_path(path: &mut PathBuf, is_sending: bool, key: &SnapKey) -> io::Result<()> {
+            if !path.exists() {
+                try!(fs::create_dir_all(path.as_path()));
+            }
+            let prefix = if is_sending {
+                SNAP_GEN_PREFIX
+            } else {
+                SNAP_REV_PREFIX
+            };
+            let file_name = format!("{}_{}.snap", prefix, key);
+            path.push(&file_name);
+            Ok(())
+        }
+
+        fn init_to_write(&mut self) -> io::Result<()> {
```

Comment:
It's minimal since that is just a private function. The point is to keep the naming for all the suffixes of interfaces `Snap::new_xxx`, `SnapMgr::get_snapshot_xxx` consistent. If `for_xxing` is preferred, let's change it.